twitter-as.js

```diff
@@ -0,0 +1,226 @@
+        reg.setSenderLocalpart("twitbot");
+        reg.addRegexPattern("users", "@twitter_.*", true);
+        reg.addRegexPattern("aliases", "#twitter_@.*", true);
+        callback(reg);
+    },
+    run: function(port, config) {
+        bridge = new Bridge({
+            homeserverUrl: "http://localhost:8008",
+            domain: "localhost",
+            registration: "twitter-registration.yaml",
+            controller: {
+                onUserQuery: userQuery,
+                onEvent: eventQuery,
+                onAliasQuery: roomQuery
+            }
+        });
+        console.log("Matrix-side listening on port %s", port);
+        //Setup twitter
+        Twitter = new MatrixTwitter(bridge, config);
+        bridge.run(port, config);
+        //Register rooms
+        bridge.loadDatabases().then(() => {
+            var roomstore = bridge.getRoomStore();
+            roomstore.getRemoteRooms({}).then((rooms) => {
+                rooms.forEach((rroom, i, a) => {
+                    if (rroom.data.extras.twitter_type == 'timeline') {
+                        roomstore.getLinkedMatrixRooms(rroom.roomId).then(function(room) {
+                            //Send the userid and roomid to the twitter stack for processing.
+                            Twitter.add_timeline(rroom.data.extras.twitter_user, room[0], rroom);
+                        });
+                    }
+                });
+            });
+        })
+    }
+}).run();
+
+function userQuery(queriedUser) {
+  return Twitter.get_user_by_id(queriedUser.localpart.substr("twitter_".length)).then( (tuser) => {
+    return uploadImageFromUrl(bridge, twitter_user.profile_image_url_https, queriedUser.getId()).then((uri) => {
+      return {
+        name: twitter_user.name + " (@" + twitter_user.screen_name + ")",
+        url: uri,
+        remote: new RemoteUser(twitter_user.id_str)
+      };
+    });
+  }).catch((error) => {
+      console.error("Couldn't find the user.");
+      console.error("Reason:", error);
+  });
+}
+
+function roomQuery(alias, aliasLocalpart) {
+    console.log(aliasLocalpart);
+    return Twitter.get_user(aliasLocalpart.substr("@twitter_".length)).then((tuser) => {
+        console.log(tuser);
+        if (tuser != null) {
+            if (!tuser.protected) {
+                return constructTimelineRoom(tuser, aliasLocalpart);
+            }
+        }
+    });
+}
+
+function eventQuery(request, context) {
+    var event = request.getData();
+    console.log(event.type);
+    if (event.type == "m.room.member") {
+        if (event.membership == "invite" && event.state_key.startsWith("@twitter_")) { //We should prolly use a regex
+            var intent = bridge.getIntent(event.state_key);
+            intent.join(event.room_id);
+
+            //Set the avatar based on the 'owners' avatar.
+            intent.getClient().getProfileInfo(event.state_key, 'avatar_url').then((url) => {
```

Comment:
According to [this](https://dev.twitter.com/overview/general/user-profile-images-and-banners), a default URL will be given so it should never be null. Again, I think this is a case of validating the API to make sure the correct content is returned. I'll add a comment because I think this could be potentially confusing.